File: src/lang/lexer.rs
use std::{process::exit};
use super::defs;
use colored::Colorize;

#[derive(Debug, PartialEq, Eq, PartialOrd, Ord)]
pub enum Token {
    NUMBER(i32),
    OPERATOR(String),
    SEMICOLON
}

#[derive(Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]
pub enum LexerState {
    DEFAULT,
    SEMICOLON,
    NUMBER,
    OPERATOR
}

pub fn tokenize(src: &String) -> Vec<Token> {
    let mut buf: String = String::new();
    let mut tokens: Vec<Token> = vec![];
    let mut state = LexerState::DEFAULT;
    let mut last_state = state.clone();

    for i in src.chars() {
        if defs::OPERATORS.contains(&i) {
            state = LexerState::OPERATOR;
        } else if i.is_digit(10) {
            state = LexerState::NUMBER;
        } else if i == ';' {
            state = LexerState::SEMICOLON;
        } else if i == ' ' || i == '\n' {
            state = LexerState::DEFAULT;
        } else {
            println!("{} unexpected token", "error:".red().bold());
            exit(1);
        }

        if last_state != state && last_state != LexerState::DEFAULT {
            match last_state {
                LexerState::DEFAULT => {},
                LexerState::NUMBER => tokens.push(Token::NUMBER(buf.clone().parse::<i32>().unwrap_or(0))),
                LexerState::OPERATOR => tokens.push(Token::OPERATOR(buf.clone())),
                LexerState::SEMICOLON => tokens.push(Token::SEMICOLON),
                _ => {}
            }

            buf = String::new();
        }

        if state == LexerState::NUMBER || state == LexerState::OPERATOR {
            buf.push(i);
        }

        last_state = state.clone();
    }

    return tokens;
}

File: src/lang/lexer.cpp
#include "lexer.hpp"
#include "defs.hpp"
#include <cctype>
#include <string>
#include <algorithm>
#include <util/error.hpp>

std::ostream& operator<<(std::ostream& os, const Token& token) {
	switch (token.type) {
		case TokenType::TOKEN_OPERATOR: return os << "OPERATOR(" << token.literal << ")";
		case TokenType::TOKEN_NUMBER: return os << "NUMBER(" << token.num << ")";
		case TokenType::TOKEN_SEMICOLON: return os << "SEMICOLON"; break;
		default: return os;
	}
}

std::ostream& operator<<(std::ostream& os, const Lexer& lexer) {
	os << "[" << std::endl;
	for (auto x : lexer.tokens) {
		os << "\t" << x << std::endl;
	}
	os << "]";
	return os;
}

Lexer::Lexer(std::string code): code(code), state(LexerState::LEXER_DEFAULT), lastState(LexerState::LEXER_DEFAULT), buffer(""), tokens() {}

Lexer::~Lexer() {}

void Lexer::tokenize() {
	for (auto c : code) {
		if (std::find(operators.begin(), operators.end(), c) != operators.end()) {
			state = LexerState::LEXER_OPERATOR;
			buffer += c;
		} else if (std::isdigit(c)) {
			state = LexerState::LEXER_NUMBER;
			buffer += c;
		} else if (c == ';') {
			state = LexerState::LEXER_DEFAULT;
			buffer = "";
			tokens.push_back(Token { TokenType::TOKEN_SEMICOLON, 0, "" });
		} else if (c == ' ' or c == '\n') {
			state = LexerState::LEXER_DEFAULT;
		} else {
			error("unexpected token");
		}

		if (lastState != state && lastState != LexerState::LEXER_DEFAULT && buffer != "") {
			switch (lastState) {
				case LexerState::LEXER_NUMBER: tokens.push_back(Token { TokenType::TOKEN_NUMBER, std::stoi(buffer), "" }); break;
				case LexerState::LEXER_OPERATOR: tokens.push_back(Token { TokenType::TOKEN_OPERATOR, 0, buffer }); break;
				default: break;
			}

			buffer = "";
		}

		lastState = state;
	}

	if (!buffer.empty()) {
		switch (lastState) {
			case LexerState::LEXER_NUMBER: tokens.push_back(Token { TokenType::TOKEN_NUMBER, std::stoi(buffer), "" }); break;
			case LexerState::LEXER_OPERATOR: tokens.push_back(Token { TokenType::TOKEN_OPERATOR, 0, buffer }); break;
			default: break;
		}

		buffer = "";
	}

}